File: Data Structures & Algorithms/max-water-container/submission-2.py
class Solution:
    def maxArea(self, h: List[int]) -> int:
        l, r = 0, len(h) - 1
        maxArea = 0

        while l < r:
            area = (r - l) * min(h[l], h[r])
            maxArea = max(maxArea, area)

            if h[l] <= h[r]:
                l += 1
            else:
                r -= 1

        return maxArea

File: Data Structures & Algorithms/eating-bananas/submission-2.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int l = 1;
        int r = piles[0];
        for (int p : piles) r = Math.max(r, p);

        int res = r;
        while (l <= r) {
            int k = (l + r) / 2;
            int total = 0;
            for (int p : piles) {
                total += (int) Math.ceil((double)p/k);
            }
            if (total > h) {
                l = k + 1;
            } else {
                res = k;
                r = k - 1;
            }
        }
        return res;
    }
}
